src/System.Net.Security/src/System/Net/Security/SslState.cs

```diff
@@ -1183,43 +1183,28 @@ private void FinishHandshakeRead(int newState)
                 }
 
                 _lockReadState = LockRead;
-                object obj = _queuedReadStateRequest;
-                if (obj == null)
-                {
-                    // Other thread did not get under the lock yet.
-                    return;
-                }
-
-                _queuedReadStateRequest = null;
-                if (obj is LazyAsyncResult)
-                {
-                    ((LazyAsyncResult)obj).InvokeCallback();
-                }
-                else
-                {
-                    ThreadPool.QueueUserWorkItem(new WaitCallback(CompleteRequestWaitCallback), obj);
```

Comment:
You are modifying this code to now do, AsyncResumeHandshake, instead of CompleteRequestWaitCallback, in the new HandleQueuedCallback. When I looked at it, this code will be hit during re-negotiation request on read/write. If you are confident this codepath will not be hit with your changes, remove the default case in the switch in the new function, so that if breaks we know we shouldn't change this.

Reply:
@Drawaes  This difference in behavior also existed in the WriteAsync PR, I had pointed it out in one of my previous PRs and you mentioned that this codepath will not be hit. Can you also verify with a re-negotiation test..

Reply:
How is this behaviour different to what was there previously?

As for test coverage well... In a word, no I brought up the lack of re-negotiation tests a while ago in the issue #24407 . Its not something you can do from .net and I don't know how to sort this out in your outerloop infra. So it will need to be done by a microsoftie, it will either need to be an "OpenSSL" server setup or IIS with certificate validation on a certain path to trigger the renegotiation.

Reply:
> How is this behaviour different to what was there previously?

The function HandleQueuedCallback is doing ```ThreadPool.QueueUserWorkItem(new WaitCallback(AsyncResumeHandshake), obj);``` instead of ```ThreadPool.QueueUserWorkItem(new WaitCallback(CompleteRequestWaitCallback), obj);``` which this code did before replacing with HandleQueuedCallback.

> As for test coverage well...

I was hinting towards testing using .netcore client sslstream, and testing with some thirdparty sslstream test server that supports re-negotiation. We need to do a sanity check to ensure this scenario is not broken.

Reply:
Yeah, I have done that locally, it works, but I can't "productionize" this as it would require doing things on the outerloop servers that I just don't know how (or think I even can) do.

Reply:
> Yeah, I have done that locally, it works

If that is the case, and the default: case in the switch statement is not hit, remove that code from the default case, since it's not reachable, and the logic doesn't make sense to be there in the ```HandleQueuedCallback``` function.

Reply:
You can't I will try to explain with a bit of code (I am happy to put it back as three methods if it is a sticking point but I find code makes it easier to understand)

Originally there were three bits of code doing this

``` csharp

// Finish Handshake Read

if (obj is LazyAsyncResult)
{
	((LazyAsyncResult)obj).InvokeCallback();
}
else
{
	ThreadPool.QueueUserWorkItem(new WaitCallback(CompleteRequestWaitCallback), obj);
}

// Finish Write

if (obj is LazyAsyncResult)
{
	// Sync handshake is waiting on other thread.
        ((LazyAsyncResult)obj).InvokeCallback();
}
else
{
	// Async handshake is pending, start it on other thread.
        // Consider: we could start it in on this thread but that will delay THIS write completion
        ThreadPool.QueueUserWorkItem(new WaitCallback(AsyncResumeHandshake), obj);
}

// FinishHandshake

if (obj is LazyAsyncResult)
{
	// Sync write is waiting on other thread.
        ((LazyAsyncResult)obj).InvokeCallback();
}
else
{
	// Async write is pending, start it on other thread.
        // Consider: we could start it in on this thread but that will delay THIS handshake completion
        ThreadPool.QueueUserWorkItem(new WaitCallback(CompleteRequestWaitCallback), obj);
}
```

Now the two completerequestwaitcallback bits of code will now have been replaced with a TaskCompletionSource as they are Async not APM. There is also code at the top I have left out for brevity but its basically a null check and null out of the "obj" which is identical.

So we can see the "Is LazyAsyncResult" is identical so it transfers directly to the switch.

The next bit, as said is a TCS for the Finish handshake read and finish handshake (because it would be the read/write that queued while they happened). So that only leaves the last Threadpool APM bit of code, which is the "Default". 

I would hope to eliminate the default case with a change to async/await for the handshake (as well as the TCS's as it should be able to just be a SemaphoreSlim for the read/write locks instead of all of the Interlocks, and extra baggage). This is something I am working on, and it massively cleans up this code (in my personal opinion). Its just a case of it had to start somewhere and this is the point that the async/await and the APM collide.